File: app/src/main/java/com/liweiyap/narradir/util/SamsungSemEmergencyManagerLeakFixActivity.java
package com.liweiyap.narradir.util;

import android.app.Activity;
import android.app.Application;
import android.os.Build;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.lang.reflect.Field;

/**
 * Fixes a (third-party) library leak in Samsung Galaxy A3 (2016). See: https://github.com/square/leakcanary/issues/762
 * Might be removed eventually if we decide to stop supporting Android 7
 */
public final class SamsungSemEmergencyManagerLeakFixActivity implements Application.ActivityLifecycleCallbacks
{
    private SamsungSemEmergencyManagerLeakFixActivity(Application application)
    {
        mApplication = application;
    }

    @Override
    public void onActivityCreated(@NonNull Activity activity, @Nullable Bundle savedInstanceState){}

    @Override
    public void onActivityStarted(@NonNull Activity activity){}

    @Override
    public void onActivityResumed(@NonNull Activity activity){}

    @Override
    public void onActivityPaused(@NonNull Activity activity){}

    @Override
    public void onActivityStopped(@NonNull Activity activity){}

    @Override
    public void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle outState){}

    @Override
    public void onActivityDestroyed(@NonNull Activity activity)
    {
        try
        {
            swapActivityWithApplicationContext();
        }
        catch (Exception e)
        {
            // The same result is expected on subsequent tries.
            e.printStackTrace();
        }

        mApplication.unregisterActivityLifecycleCallbacks(this);
    }

    public static void applyFix(Application application)
    {
        if ( (Build.MANUFACTURER.equals("samsung")) && (Build.VERSION.SDK_INT <= Build.VERSION_CODES.N) )
        {
            application.registerActivityLifecycleCallbacks(new SamsungSemEmergencyManagerLeakFixActivity(application));
        }
    }

    private void swapActivityWithApplicationContext() throws ClassNotFoundException, NoSuchFieldException, IllegalAccessException
    {
        Class<?> semEmergencyManagerClass = Class.forName("com.samsung.android.emergencymode.SemEmergencyManager");
        Field sInstanceField = semEmergencyManagerClass.getDeclaredField("sInstance");
        sInstanceField.setAccessible(true);
        Object sInstance = sInstanceField.get(null);
        Field mContextField = semEmergencyManagerClass.getDeclaredField("mContext");
        mContextField.setAccessible(true);
        mContextField.set(sInstance, mApplication);
    }

    private final Application mApplication;
}

File: app/src/main/java/com/liweiyap/narradir/util/SamsungSemEmergencyManagerLeakFixActivity.kt
package com.liweiyap.narradir.util

import android.app.Activity
import android.app.Application
import android.os.Build
import android.os.Bundle

import java.lang.reflect.Field

/**
 * Fixes a (third-party) library leak in Samsung Galaxy A3 (2016). See: https://github.com/square/leakcanary/issues/762
 * Might be removed eventually if we decide to stop supporting Android 7
 */
class SamsungSemEmergencyManagerLeakFixActivity private constructor(private val mApplication: Application): Application.ActivityLifecycleCallbacks {
    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}
    override fun onActivityStarted(activity: Activity) {}
    override fun onActivityResumed(activity: Activity) {}
    override fun onActivityPaused(activity: Activity) {}
    override fun onActivityStopped(activity: Activity) {}
    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
    override fun onActivityDestroyed(activity: Activity) {
        try {
            swapActivityWithApplicationContext()
        }
        catch (e: Exception) {
            // The same result is expected on subsequent tries.
            e.printStackTrace()
        }

        mApplication.unregisterActivityLifecycleCallbacks(this)
    }

    @Throws(
        ClassNotFoundException::class,
        NoSuchFieldException::class,
        IllegalAccessException::class)
    private fun swapActivityWithApplicationContext() {
        val semEmergencyManagerClass: Class<*> = Class.forName("com.samsung.android.emergencymode.SemEmergencyManager")
        val sInstanceField: Field = semEmergencyManagerClass.getDeclaredField("sInstance")
        sInstanceField.isAccessible = true
        val sInstance: Any? = sInstanceField.get(null)
        val mContextField: Field = semEmergencyManagerClass.getDeclaredField("mContext")
        mContextField.isAccessible = true
        mContextField.set(sInstance, mApplication)
    }

    companion object {
        @JvmStatic
        fun applyFix(application: Application) {
            if ( (Build.MANUFACTURER == "samsung") && (Build.VERSION.SDK_INT <= Build.VERSION_CODES.N) ) {
                application.registerActivityLifecycleCallbacks(SamsungSemEmergencyManagerLeakFixActivity(application))
            }
        }
    }
}